EVG-11867 | Product Recommendations Rows

amaskiellEG opened the pull request:
# Description

Adds the Global Template for Product Recommendations Rows (using [NTO](https://www.northerntrailoutfitters.com/default/homepage))

This Global Template inserts a row of recommended Products, chosen by a selected Einstein Recipe.

## Related Issues/Tickets

JIRA: https://help.evergage.com/browse/EVG-11867

## How to test

#### Template: 
- You can view the demo template in eng1:globaltemplates, **template name: [EVG-11867] Global - Product Recommendations Rows**. 

#### Campaign: 
- You can view the demo campaign using this template in eng1:globaltemplates, **campaign name: [EVG-11867] Global - Product Recommendations Rows**
- Test Links with various layouts: 
-- The Works: https://www.northerntrailoutfitters.com/default/homepage?evergageTestMessages=gcHBe
-- No Name: https://www.northerntrailoutfitters.com/default/homepage?evergageTestMessages=0QXKc
-- No Description: https://www.northerntrailoutfitters.com/default/homepage?evergageTestMessages=6K3Wx
-- No Price: https://www.northerntrailoutfitters.com/default/homepage?evergageTestMessages=uRwlW
-- No Rating: https://www.northerntrailoutfitters.com/default/homepage?evergageTestMessages=BymmS
-- Name + Description: https://www.northerntrailoutfitters.com/default/homepage?evergageTestMessages=eZq3d
-- Name + Price: https://www.northerntrailoutfitters.com/default/homepage?evergageTestMessages=5Ms3V
-- Price Only: https://www.northerntrailoutfitters.com/default/homepage?evergageTestMessages=zy5fp

_Note: If you test on [NTO](https://northerntrailoutfitters.com/default/homepage), you will have to inject the Beacon._

### Things to check for:
* **Visuals:** Does it look good across different client sites? To test this, you can copy, paste, and run this snippet in the JS console: https://gist.github.com/amaskiellEG/be83320023b2ada067cb0f60f5516364 (Note that it will appear at the very top of the page.)
* **Mobile/smaller screen responsiveness:** Does the Template generally look good on smaller screens?
* **WCAG compliant:** Does the template contain ARIA attributes and/or use semantic elements for web accessibility? 
** You can use the WAVE Chrome extension tool https://chrome.google.com/webstore/detail/wave-evaluation-tool/jbbplnpkjmmeebjpijfedlgcdilocofh for a quick check. 
* **Template-specific documentation**: Is there sufficient information for a Template Developer to use the Global Template and customize it for their own use?
------------ ONLY TESTABLE ON NTO -------------
* **Deferred rendering:** Is the rendering of the Template deferred until the Content Zone element is loaded on page? 
* **Flicker defense:** Is the Template flicker-defended?
* **Campaign stats tracking:** Do impression stats and click stats fire for both the Control and Test experience?

**Important note:** In order to test **Flicker Defense** properly, we want to test the template on a site that already loads Beacon v16. The [NTO](https://northerntrailoutfitters.com/default/homepage) demo site does _not_ have the Beacon loaded yet and it’s not clear when it’ll be added. With the Beacon on the Dev Docs site, it's possible to test this there. Although not as easy to use for testing, another option is to capture the page load with Chrome's performance option and then look at the changes in frame in relation to the Beacon's injection.

amaskiellEG commented on product-recs/template.css:
```diff
@@ -1,47 +1,157 @@
+/* ==========================================================================
+    EVG Global Styles
+========================================================================== */
+
+.evg-h1,
+.evg-h2,
+.evg-h3,
+.evg-h4,
+.evg-h5,
+.evg-h6 {
+    padding: 0;
+    margin: 0 0 .5rem;
+    font-family: inherit;
+    font-weight: 500;
+    line-height: 1.2;
+    color: inherit;
+}
+
+.evg-h3 {
+    font-size: 1.75rem;
+}
+
```
Thought it might be good to have these as separate, just so that they're consistent with what's in https://confluence.internal.salesforce.com/pages/viewpage.action?pageId=245293245 but I'm not entirely sold on it. Since there's only using one header be used (`h3`), it might just make sense to put these all under `#evg-product-recs .evg-h3`?

amaskiellEG commented on product-recs/template.hbs:
```diff
@@ -1,18 +1,44 @@
+            <a href="{{attributes.url.value}}">
+                <img class="evg-product-rec-image" src="{{attributes.imageUrl.value}}" data-evg-product-imageUrl="{{attributes.imageUrl.value}}" alt="{{attributes.name.value}}"/>
+            </a>
+            <div class="evg-product-rec-text" aria-label="Product Text for {{attributes.name.value}}">
+                <a class="evg-product-rec-name" href="{{attributes.url.value}}" data-evg-product-name="{{attributes.name.value}}" aria-label="Name: {{attributes.name.value}}">
+                    {{attributes.name.value}}
+                </a>
+                <div class="evg-product-rec-price" data-evg-product-price="{{attributes.price.value}}" aria-label="Price: {{attributes.price.value}}">
+                    ${{attributes.price.value}}
+                </div>
+                <div class="evg-product-rec-rating" data-evg-product-rating="{{attributes.rating.value}}" aria-label="Rating: {{attributes.rating.value}}">
```
Are these `aria-label`s overkill? Didn't know if it'd be better to play it safe and add them.

amaskiellEG commented on product-recs/template.ts:
```diff
@@ -1,23 +1,75 @@
+    /**
+     * Business-User Controls
+     */
+    @options([
+        { name: "home_product_recommendations", label: "Homepage Product Recommendations" },
+        { name: "product_detail_recs_row_1", label: "Product Details Recommendations Row 1" },
+        { name: "product_detail_recs_row_2", label: "Product Details Recommendations Row 2" },
+        // TODO: Needed in sitemap
+        // { name: "plp_recs_row", label: "PLP Recommendations Row" },
+        // { name: "cart_recs_row_1", label: "Cart Recommendations Row 1" },
+        // { name: "cart_recs_row_2", label: "Cart Recommendations Row 2" },
+        // { name: "order_confirmation_recs_row", label: "Order Confirmation Recommendations Row" },
+        // { name: "arc_modal_recs_row", label: "Arc Modal Recommendations Row" },
+        // { name: "404_error_recs_row", label: "404 Error Recommendations Row" },
+        // { name: "search_results_recs_row", label: "Search Results Recommendations Row" },
+    ])
+    selectedContentZone: ProductRecommendationsContentZones = { name: "home_product_recommendations", label: "Homepage Product Recommendations" };
+
+    @hidden(true)
+    contentZone: string = this.selectedContentZone.name;
```
I think all of this is only necessary if different labels are wanted for the business-user input.

If that's not the case, one of these can replace this:
_Option 1_
```
options([
    "home_product_recommendations",
    "product_detail_recs_row_1",
    "product_detail_recs_row_2"
])
contentZone: string = "home_product_recommendations";
```

_Option 2_
```
contentZone: "home_product_recommendations" | "product_detail_recs_row_1" | "product_detail_recs_row_2" = "home_product_recommendations";
```

(FWIW, I think Option 1 is a little cleaner and easier to read.)

avkieu replied:
Do not use labels in the template for content zones. And not because you have to do this weird extra work to actually make it work but because, not only do we use the field `contentZone` to derive where a configured campaign shows, we use the field to determine all of the eligible zones for any given template. This means, the options MUST exist on `contentZone`.

This is used to generate summaries about the contentZones, so that in the future, you'll be able to click a highlighted contentZone in the VE (labeled with the contentZoneId!) and say, lets take this zone and this template and start a campaign.

Some other less egregious reasons.

1. It's creating this standard people are going to look at, and apply in their templates, but there's nothing to ensure everyone uses the same label for the same zones across all the templates. So now you have zones, that in the eyes of the Business User, have different names for the same things.

2. The campaign editor actually has some UI in it that references the content zones, by their content zone id. This makes the field in their config NOT match the field in the UI.

3. The content zones, by IDs, are going to be present in a lot of places, visible to business users. For instance, we'll be putting borders around zones on their site in the VE, labeled with their ID, and some action you can take, like create a new campaign on this zone. The IDs need to be recognizable by business users, and in a way that is a shared language with the developers.

amaskiellEG commented on product-recs/template.ts:
```diff
@@ -1,23 +1,75 @@
+    recsConfig: RecommendationsConfig = new RecommendationsConfig().restrictItemType("Product").restrictMaxResults(this.maxResults)
+
+    @title("Show Product name")
+    // @shownIf()
+    nameVisibility: boolean = true;
+
+    @title("Show Product description")
+    // @shownIf()
+    descriptionVisibility: boolean = true;
+
+    @title("Show Product price")
+    // @shownIf()
+    priceVisibility: boolean = true;
+
+    @title("Show Product rating")
+    // @shownIf()
```
Waiting for clarity about `shownIf` settings and using `description` at all

amaskiellEG replied:
Got confirmation from @nriserEG that `@shownIf` settings can be removed.

nriserEG commented on product-recs/template.css:
```diff
@@ -1,47 +1,155 @@
+   Media Queries
+========================================================================== */
+
+@media only screen and (max-width: 990px) {
```
Use 991px

Also, no need for "only" here because we don't support browsers that don't work with media queries anyway (https://caniuse.com/#feat=css-mediaqueries)

nriserEG commented on product-recs/template.ts:
```diff
@@ -1,23 +1,71 @@
-            validate: () => true
-        });
+    @title(" ")
+    recsConfig: RecommendationsConfig = new RecommendationsConfig().restrictItemType("Product").restrictMaxResults(this.maxResults)
```
Spoke with @aspatti1257 about this. Remove `.restrictMaxResults(this.maxResults)` here and add this line in the `run` function:
```
        this.recsConfig.maxResults = this.maxResults;
```
With this setup, coupled with the separate `maxResults` config, `.restrictMaxResults(this.maxResults)` no longer makes sense to have (and is rather confusing).

amaskiellEG replied:
I was using `restrictMaxResults` based on the README in the `recs` Gear: https://github.com/evergage/evergage-product/tree/master/analytics/excavator/src/main/resources/built-in/gears/recs

`maxResults` is only meant to be changed by a developer, but the reason I created that setting in that format is based on the request doc, between the wording of "template can accommodate 4-6 products" and the `numberOfProducts` setting under the "hidden controls" section:

https://docs.google.com/document/d/1BpFoQ7MFviuxha5ld7LM2vfuJ-sPAaAXgdjVSOdadrY/edit#

I originally thought that it could be a value changed by the business user, but then realized that seemed contradicted by the note under "hidden controls".

Aside from me misnaming `numberOfProducts` as `maxResults`, it seemed like it made the most sense to wrap those considerations into one value and then pass the reference to that value to `restrictMaxResults`. If we don't need that setting at all, then might be easier to just eliminate it and just use `.resrictMaxResults(6)` instead, which would also be more consistent with the documentation in the `recs` Gear.

nriserEG replied:
Yes, we'd like to keep the separate maxResults that you made because that's what's requested in the Requirements doc. To accommodate this, we simply need to remove the `.restrictMaxResults(...)` and add `this.recsConfig.maxResults = this.maxResults;` (There's a separate PR comment -- line 66) to make it all work.

amaskiellEG replied:
Should `maxResults` be changed to `numberOfProducts`, since that's what's in the request?

nriserEG replied:
@amaskiellEG Yes let's do that! 👍

nriserEG replied:
As we spoke on Slack, looks like we'll need `.restrictMaxResults(this.maxResults)` after all to avoid having the default max results config show up!

aspatti1257 replied:
Just chiming in here:
`numberOfProducts` implies that it will always show the number given. If you set it to 4, people may expect that 4 will return all the time, when in reality, it may be only 2. `maxResults` implies the number is <= to this value. Which is more accurate as to what is actually going on.

nriserEG replied:
thanks @aspatti1257, good call! @amaskiellEG Let's change it to maximumNumberOfProducts. I'll let Kyle know of the change 👍

nriserEG replied:
Checked with Kyle - he's cool with it 👍

nriserEG commented on product-recs/template.ts:
```diff
@@ -1,23 +1,71 @@
+            price: this.priceVisibility,
+            rating: this.ratingVisibility
+        };
+
```
Add this line: 
```
        this.recsConfig.maxResults = this.maxResults;
```

amaskiellEG replied:
Going to include this because of breaking "default" config changes (related to: https://evergage.slack.com/archives/CJDLXS8MT/p1597099979051200)

nriserEG commented on product-recs/template.ts:
```diff
@@ -1,23 +1,71 @@
+     * Developer Controls
+     */
+    @hidden(true)
+    maxResults: 4 | 5 | 6 = 6;
```
Spoke with Kyle about this. We'd like the numbers to be all even. Update to 2, 4, 6, and 8

nriserEG commented on product-recs/client.js:
```diff
@@ -1,22 +1,95 @@
+         * Visit the Campaign Stats Tracking documentation to learn more:
+         * https://developer.evergage.com/templates/campaign-stats
+         */
+        var contentZoneSelector = Evergage.getContentZoneSelector(context.contentZone);
+        Evergage.cashDom(contentZoneSelector).attr("data-evg-campaign-id", context.campaign);
+        Evergage.cashDom(contentZoneSelector).attr("data-evg-experience-id", context.experience);
+        Evergage.cashDom(contentZoneSelector).attr("data-evg-user-group", "Control");
+        Evergage.cashDom(contentZoneSelector + " a").attr("data-evg-clickthrough", "");
```
I don't think we need control tracking, since this template is meant for _inserting_ recs. However, maybe we could add a note in the Handlebars comment (under "Customizations") that they could add control tracking if they intend on using this template to _replace_ their recs (and direct them to the stats doc). 

Thoughts @avkieu @dcohenEG?

avkieu replied:
I think it's fine to leave it, as it won't do anything bad when inserting, and will just work when replacing.

If it's "inserts" or "replaces" is really dependent on the contentZone and the sitemap, not the template in this case.

nriserEG commented on product-recs/client.js:
```diff
@@ -1,22 +1,95 @@
+         * with the HTML generated from Handlebars.
+         *
+         * If you instead wish to insert your generated HTML before or after your
+         * selected content zone, use .before(html) or .after(html), respectively.
```
Update this comment to include the option of replacing the content.

nriserEG commented on product-recs/client.js:
```diff
@@ -1,22 +1,95 @@
+                Evergage.cashDom(contentZoneSelector).before(html);
+                applyVisibilityOptions(context);
+            }
+        });
```
- Change `el` to `element`
- use `element` instead of `contentZoneSelector` to append `html`
- Remove `if` conditional
```
        var contentZoneSelector = Evergage.getContentZoneSelector(context.contentZone);
        return Evergage.DisplayUtils.pageElementLoaded(contentZoneSelector).then(element => {
            var html = template(context);
            Evergage.cashDom(element).before(html);
            applyVisibilityOptions(context);
        });
```

nriserEG commented on product-recs/template.hbs:
```diff
@@ -1,18 +1,44 @@
+    5) Change any and all values of aria-label attributes as necessary.
+--}}
+
+<div id="evg-product-recs" data-evg-campaign-id="{{campaign}}" data-evg-experience-id="{{experience}}" data-evg-campaign-group="test" aria-label="Product Recommendations {{#if header}}- {{header}}{{/if}}">
```
Change `data-evg-campaign-group="test"` to `data-evg-user-group="{{userGroup}}"`

amaskiellEG commented on product-recs/template.ts:
```diff
@@ -1,23 +1,75 @@
+    /**
+     * Business-User Controls
+     */
+    @options([
```
Is that necessary since the title is generated already?

nriserEG replied:
huh, I thought I saw a different config name -- yes, please ignore!

amaskiellEG commented on product-recs/template.ts:
```diff
@@ -1,23 +1,75 @@
+    contentZone: string = this.selectedContentZone.name;
+
     @title("Recommendations Row Header")
     header: string;
```
I'll change it but it _does_ seem a little odd to have something with directly the same name as a decorator?

```
@title("Recommendations Row Header")
title: string = "You May Also Like";
```

nriserEG replied:
This is what I meant: 
```
    @subtitle("i.e. You May Also Like")
    title: string = "You May Also Like";
```

<img width="344" alt="Screen Shot 2020-08-10 at 4 54 29 PM" src="https://user-images.githubusercontent.com/32201252/89842375-49b61300-db2a-11ea-916b-47857752bb1a.png">

Requirements doc: 
<img width="226" alt="Screen Shot 2020-08-10 at 4 56 26 PM" src="https://user-images.githubusercontent.com/32201252/89842427-6fdbb300-db2a-11ea-8d6c-8f01534a538f.png">

amaskiellEG replied:
Ah ok! Sorry that I somehow missed that.

It still seems a little weird to me to have config named after a decorator:
```
@title("Title")
@subtitle("i.e. You May Also Like")
title: string = "You May Also Like";
```

Maybe it'd make more sense to the developer if it was called `header`?
```
@title("Title")
@subtitle("i.e. You May Also Like")
header: string = "You May Also Like";
```

nriserEG replied:
Sure, sounds good 👍  I'll make the update in the Content Recs global template too

nriserEG commented on product-recs/template.ts:
```diff
@@ -1,23 +1,57 @@
-            recipeId: this.recipeId,
-            validate: () => true
-        });
+    title: string = "You May Also Like";
```
Change to 
```
    @subtitle("i.e. You May Also Like")
    title: string = "You May Also Like";
```

<img width="344" alt="Screen Shot 2020-08-10 at 4 54 29 PM" src="https://user-images.githubusercontent.com/32201252/89842375-49b61300-db2a-11ea-916b-47857752bb1a.png">

Requirements doc: 
<img width="226" alt="Screen Shot 2020-08-10 at 4 56 26 PM" src="https://user-images.githubusercontent.com/32201252/89842427-6fdbb300-db2a-11ea-8d6c-8f01534a538f.png">

nriserEG commented on product-recs/template.ts:
```diff
@@ -1,23 +1,57 @@
-        });
+    title: string = "You May Also Like";
+
+    @title(" ")
```
add @subtitle("i.e. Collaborative with Trending") 
(from the Requirements doc)

Note: I have this in my Content Recommendations global template as well, and when I added it, I notice that it appears after an empty extra line under the Recipe dropdown. Example (compare the subtitle from the Title config to that of the Recipe config): 
<img width="338" alt="Screen Shot 2020-08-10 at 5 08 23 PM" src="https://user-images.githubusercontent.com/32201252/89842988-1aa0a100-db2c-11ea-868c-4cc827654ec2.png">
Minor issue, but we'll want to remove that empty line at one point.
@aspatti1257 I'm guessing that this has to do with the Recs gear, but is this something that could be fixed?

nriserEG commented on product-recs/client.js:
```diff
@@ -1,22 +1,72 @@
\ No newline at end of file
+    function reset(context, template) {
+        /** Remove the template from the DOM to reset the template. */
+        Evergage.cashDom(buildExperienceSelector(context)).remove();
```
Change to this for simplicity:
```
Evergage.cashDom("#evg-recipe-product-recommendations").remove();
```

amaskiellEG replied:
As long as we're staying consistent with our naming of `id`s, that's ok, but I wonder if it makes sense to abstract it into a variable? 

(I also realize now that the helper function I had is excessive/unnecessary at this point 😅 )

nriserEG replied:
Yes this is consistent with best practices defined in https://developer.evergage.com/templates/templates-style-guide-and-coding-conventions 
> Add an ID attribute to the outermost element, using a name that is specific to the template that you are building, and prefix with the evg-namespace (e.g. id=“evg-hp-hero-banner”). Use an ID for only this outermost HTML element and do not use the same ID in more than one template.
Let's stick with this

amaskiellEG replied:
I guess I had something like this in mind:
```
(function() {

    const templateId = "#evg-recipe-product-recommendations";

    ...

    function reset(context, template) {
        Evergage.cashDom(templateId).remove();
    }

    ...

})();
```

In this case, it's not really necessary since it's only referenced once, but I'm wondering if it might be a chance to set the standard for people to get into the habit setting strings as variables when it's likely that they'll be reused? I'm happy to also just do what you're suggesting, but wanted to present the idea. 🙂

amaskiellEG replied:
Talked to @nriserEG and we decided it's best to just use `Evergage.cashDom("#evg-recipe-product-recommendations").remove();`

One thing to note is that it will _only_ target and remove the _first_ element with that `id`, meaning that in rare occurrences when multiple elements with that `id` might be present, they _all_ will not be removed when `reset` is run.

nriserEG commented on product-recs/client.js:
```diff
@@ -1,22 +1,72 @@
-})();
\ No newline at end of file
+    function reset(context, template) {
+        /** Remove the template from the DOM to reset the template. */
```
Remove comment as this is not template specific (See Hero Banner template)

nriserEG commented on product-recs/client.js:
```diff
@@ -1,22 +1,72 @@
-    var html = template(context);
-    Evergage.cashDom(".experience-component.experience-layouts-1_column:nth-of-type(9)").after(html);
-  }
+    /**
+     * @function buildExperienceSelector
+     * @param {Object} context
+     * @description Concatenates selector using Experience ID from context.
+     */
+    function buildExperienceSelector(context) {
+        return `[data-evg-experience-id=${context.experience}]`;
+    }
```
Remove lines 3-10, since we simply want to refer to the template ID selector later in the code

nriserEG commented on product-recs/client.js:
```diff
@@ -1,22 +1,72 @@
-  }
+    function apply(context, template) {
+        /**
+         * The code below insert your generated HTML from Handlebars before the
+         * selected content zone by using .before(html), though .after(html) can
+         * be used if you'd like insert the HTML after that content zone.
+         *
+         * If you instead wish to replace the content of your selected content zone
+         * with the HTML generated from Handlebars, then .html(html) can be used.
+         *
+         * Note: To use a content zone, you must define a "contentZone" configuration
+         * property in the Serverside Code.
+         *
+         * Use Evergage.DisplayUtils.pageElementLoaded to defer the rendering of
+         * the template until the content zone element is loaded on page.
+         * The observer element that monitors for the content zone element to get inserted
+         * into its DOM node is set to "body" by default. For performance optimization, this
+         * default can be overridden by adding a second selector argument, which will be used
+         * as the observer element instead.
+         *
+         * Visit the Display Utilities documentation to learn more:
+         * https://developer.evergage.com/templates/display-utilities
+         *
+         * Note: To use Evergage.DisplayUtils.pageElementLoaded, you must have the
+         * Display Utilities gear installed and enabled for your dataset.
```
Remove lines 33-35. We need to temporarily remove any comment having to do with installing a gear

nriserEG replied:
Remove lines 14-35, as it's too broad. Follow the example in the Hero Banner template -- copy the comment here: 
https://github.com/evergage/evergage-global-templates/blob/master/hero-banner-cta/client.js

nriserEG commented on product-recs/client.js:
```diff
@@ -1,22 +1,72 @@
+    }
+
+    function control(context) {
+        /**
+         * Add Evergage data attributes to elements you wish to track in the control experience.
+         *
+         * Note: Using Evergage data attributes allows you to track stats with Evergage and
+         * requires the Campaign Stat Tracking gear installed and enabled for your dataset.
+         * Visit the Campaign Stats Tracking documentation to learn more:
+         * https://developer.evergage.com/templates/campaign-stats
+         */
```
Remove comment, lines 51-58, since it's not template specific. See Hero Banner template

nriserEG commented on product-recs/template.hbs:
```diff
@@ -1,18 +1,58 @@
-            <div class="evg-price">
-                ${{attributes.price.value}}
+{{!--
+    Global Template: Product Recommendations Rows
```
Change to the official template name: "Recipe Product Recommendations"

nriserEG commented on product-recs/client.js:
```diff
@@ -1,22 +1,72 @@
+        Evergage.cashDom(contentZoneSelector).attr("data-evg-campaign-id", context.campaign);
+        Evergage.cashDom(contentZoneSelector).attr("data-evg-experience-id", context.experience);
+        Evergage.cashDom(contentZoneSelector).attr("data-evg-user-group", "Control");
+        Evergage.cashDom(contentZoneSelector + " a").attr("data-evg-clickthrough", "");
```
Let's switch to template literals syntax
```
Evergage.cashDom(`${contentZoneSelector} a`).attr("data-evg-clickthrough", "");
```

nriserEG commented on product-recs/template.css:
```diff
@@ -1,47 +1,177 @@
-    white-space: nowrap;
     width: 100%;
+    color: #565656;
+    font-family: Source Sans Pro, sans-serif;
```
Inherit `color` and `font-family`
```
color: inherit;
font-family: inherit;
```

nriserEG commented on product-recs/template.css:
```diff
@@ -1,47 +1,177 @@
-#evg-product-recs .evg-description {
-    padding: 10px 0px 20px;
+#evg-recipe-product-recommendations .evg-product-details {
+    padding: 1em 0;
```
use `rem`

nriserEG commented on product-recs/template.css:
```diff
@@ -1,47 +1,177 @@
+========================================================================== */
+
+#evg-recipe-product-recommendations .evg-h3 {
+    padding: 10px 0;
```
I would remove the 10px padding -- unnecessary styling, as the template looks fine without it

nriserEG commented on product-recs/template.css:
```diff
@@ -1,47 +1,177 @@
+    font-size: 1.75rem;
+    line-height: 1.2;
+    text-align: center;
 }
```
Split into two rulesets -- one for `#evg-recipe-product-recommendations .evg-h3 ` and other for `#evg-recipe-product-recommendations .evg-header`. 

Example from content recs:
```
#evg-recipe-content-recommendations .evg-h3 {
    padding: 0;
    margin: 0 0 .5rem;
    color: inherit;
    font-family: inherit;
    font-size: 1.75rem;
    font-weight: 500;
    line-height: 1.2;
}

#evg-recipe-content-recommendations .evg-header {
    margin: 1rem 0 2rem;
    width: 100%;
    font-size: 2rem;
    text-align: center;
}
```

amaskiellEG replied:
https://github.com/evergage/evergage-global-templates/pull/3#discussion_r467617581 Sorry, I'm not sure that was clear here; it seemed like you had wanted them all in one ruleset.

nriserEG commented on product-recs/template.css:
```diff
@@ -1,47 +1,177 @@
-    display: inline-block;
+
+/* ==========================================================================
+   Product Recs
```
Change to "Product Recommendations"

nriserEG commented on product-recs/evergage-template.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "name" : "Product Recommendations",
+  "name" : "Product Recommendations Rows",
```
Change to official template name - "Recipe Product Recommendations"

amaskiellEG commented on product-recs/client.js:
```diff
@@ -1,22 +1,95 @@
+        var experienceContainer = Evergage.cashDom(buildExperienceSelector(context));
+        var options = context.visibilityOptions;
+        if (typeof options === "object") {
+            Object.keys(options).map(optionKey => {
```
Not sure why I used `.map`... 😅  Good call!

However, I'm going to use `for...in`, since it's more performant than `.forEach`: https://stackoverflow.com/questions/50844095/should-one-use-for-of-or-foreach-when-iterating-through-an-array

nriserEG commented on product-recs/template.hbs:
```diff
@@ -1,18 +1,58 @@
+--}}
+
+<div id="evg-recipe-product-recommendations" data-evg-campaign-id="{{campaign}}" data-evg-experience-id="{{experience}}"
+data-evg-campaign-group="{{userGroup}}" aria-label="Product Recommendations {{#if title}}- {{title}}{{/if}}">
```
Change `data-evg-campaign-group` to `data-evg-user-group`

nriserEG commented on product-recs/template.hbs:
```diff
@@ -1,18 +1,58 @@
+    </h3>
+    <div class="evg-container evg-product-recommendations">
+        {{#each products}}
+        <div class="evg-product-recommendation" data-evg-product-id="{{id}}">
```
Remove `data-evg-product-id`
Add: `data-evg-item-id="{{id}}"` and `data-evg-item-type="{{../recipe.itemType}}"`
https://developer.evergage.com/templates/campaign-stats

nriserEG commented on product-recs/template.hbs:
```diff
@@ -1,18 +1,58 @@
+                    eg-ratings="{{#if attributes.rating.value}}{{attributes.rating.value}}{{else}}0{{/if}}">★★★★★</div>
+                    <div style="width: calc({{#if attributes.rating.value}}{{attributes.rating.value}}{{else}}0{{/if}} / 5 * 100%);"
+                    class="evg-star-fill">★★★★★</div>
+                </div>
```
I tried referencing variables in the parent scope within `#each` in the Content Recs global template and ended up not needing `this` at all. Try removing `this` from `{{#unless ../this.nameVisibility}} evg-hide{{/unless}}`, etc. 

Also, add a space before the `unless` statements. Example: 
`<div class="evg-content-name evg-card-title {{#unless ../nameVisibility}} evg-hide{{/unless}}">`

amaskiellEG replied:
Ah yeah, I found that was the case with not needing `this` too, but I just hadn't pushed up the changes yet 🙂 

Do we need that extra space? It's not necessarily going to break anything but it'll add an extra space, rendering:
`<div class="evg-content-name evg-card-title ">`
and
`<div class="evg-content-name evg-card-title  evg-hide">`

instead of
`<div class="evg-content-name evg-card-title">`
and
`<div class="evg-content-name evg-card-title evg-hide">`

nriserEG replied:
Nice! Thanks for the update!

Hmm on second thought, how about we go with:
```
<div class="evg-content-name evg-card-title {{#unless ../nameVisibility}}evg-hide{{/unless}}">
```
(I removed the space before `evg-hide`)

This will come out clean:
![Screen Shot 2020-08-11 at 1 09 43 PM](https://user-images.githubusercontent.com/32201252/89943967-eda7c900-dbd3-11ea-9c1b-72ae6a20dd77.png)

nriserEG replied:
Nice! Thanks for the update!

Hmm I was thinking it would look nicer and less confusing in the Handlebars code, but you're right, I forgot about the extra space that it would add in the rendered code. Let's keep it how you had it! 👍

nriserEG commented on product-recs/template.ts:
```diff
@@ -1,23 +1,59 @@
+    title: string = "You May Also Like";
+
+    @title(" ")
+    recsConfig: RecommendationsConfig = new RecommendationsConfig().restrictItemType("Product").restrictMaxResults(this.maximumNumberOfProducts);
```
Change the config name to `recipe` so that in Handlebars, you can reference it with `data-evg-item-type="{{../recipe.itemType}}"` (To be consistent with the recs example in Dev Docs https://developer.evergage.com/templates/campaign-stats)

nriserEG replied:
Also, to keep it under 120 characters for readability, break the line to: 
```
    recsConfig: RecommendationsConfig = new RecommendationsConfig().restrictItemType("Product")
    .restrictMaxResults(this.maximumNumberOfProducts);
```

amaskiellEG replied:
I think it might make more sense to change the documentation in this case?

It's going to conflict with documentation either way, since the example in `recs` Gear uses `recsConfig`: https://github.com/evergage/evergage-product/tree/master/analytics/excavator/src/main/resources/built-in/gears/recs

Regardless of name, I think we'd have to change it anyway in the dev docs, since it seems like what we stumbled on with scoping means that the example in the dev docs wouldn't work either (it should be `{{../recipe.itemType}}` instead of `{{recipe.itemType}}`)

I can submit a PR to change that in the dev docs, if that's ok with you?

nriserEG commented on product-recs/template.ts:
```diff
@@ -1,23 +1,65 @@
+     * Business-User Controls
+     */
+    @options([
+        "home_product_recommendations",
+        "product_detail_recs_row_1",
+        "product_detail_recs_row_2",
+        "plp_recs_row",
+        "cart_recs_row_1",
+        "cart_recs_row_2",
+        "order_confirmation_recs_row",
+        "arc_modal_recs_row",
+        "404_error_recs_row",
+        "search_results_recs_row",
```
So Kevin came up with standardized names for some of these so you'll want to update them based on this newly updated doc: https://github.com/evergage/evergage-developer-docs-static/blob/master/src/docs/websdk/contentzones.md

We're still waiting on names for the 404 error page, ATC modal page, and the order confirmation page so we will want to leave those ones out from the options list for now. We'll add them in later in the next update/version of the template: https://help.evergage.com/browse/EVG-11983

nriserEG commented on product-recs/template.ts:
```diff
@@ -1,23 +1,65 @@
-import { RecipeReference, RecipeReferenceLookup, recommend } from "recs";
+import { RecommendationsConfig, recommend } from "recs";
+
+export class ProductRecommendationsContentZones {
+    name: string;
+    label: string;
+}
```
Remnant of your old implementation -- let's remove

amaskiellEG replied:
🧟 (couldn't find a mummy one 😂 )

nriserEG commented on product-recs/template.hbs:
```diff
@@ -1,18 +1,63 @@
+    </h3>
+    <div class="evg-container evg-product-recommendations">
+        {{#each products}}
+        <div class="evg-product-recommendation" data-evg-item-id="{{id}}" data-evg-item-type="{{itemType}}">
```
We need to reach the parent scope here too as the attribute comes out empty -- add `../`

`data-evg-item-type="{{../itemType}}"`

amaskiellEG replied:
Ooops! Thanks, good catch 🙂

nriserEG commented on product-recs/client.js:
```diff
@@ -1,22 +1,49 @@
+         */
+        const contentZoneSelector = Evergage.getContentZoneSelector(context.contentZone);
+        return Evergage.DisplayUtils.pageElementLoaded(contentZoneSelector)
+            .then(element => {
```
I thought about this formatting more and I think we should stick with keeping it in the same line as line 24:

```
return Evergage.DisplayUtils.pageElementLoaded(contentZoneSelector).then(element => {
    ...
});
```

Reasons:
- Keeps nested code less indented. Makes a big difference in readability when you have a lot of code. 
- Consistency with best practices in docs and boilerplate code

amaskiellEG replied:
I think those are definitely good reasons to change it and I'm happy to do so.

I _do_ think that's worth considering a general practice I've seen often and the readability when chaining additional calls like `catch`:
```
return Evergage.DisplayUtils.pageElementLoaded(contentZoneSelector)
    .then(element => {
        const html = template(context);
        Evergage.cashDom(element).before(html);
    })
    .catch(error => {
        console.warn(error);
    });
```

amaskiellEG commented on product-recs/template.css:
```diff
@@ -1,47 +1,212 @@
+    #evg-recipe-product-recommendations .evg-product-recommendation {
+        padding: 1rem 0;
+        margin-bottom: 2rem;
+        max-width: calc(50% - 1rem)!important;
```
In this case, yes, because it needs to override an inline style. I'll add that space 🙂

nriserEG replied:
makes sense!

amaskiellEG commented on product-recs/template.css:
```diff
@@ -1,47 +1,212 @@
+    width: 100%;
+}
+
+#evg-recipe-product-recommendations .evg-product-name--wrapper {
+    margin-bottom: .25rem;
+    display: flex;
+    min-height: 3rem;
+}
+
+#evg-recipe-product-recommendations .evg-product-name {
+    margin: auto auto 0;
+    min-height: 2.5rem;
+    max-width: 90%;
+    display: block;
+    color:#393939;
+    font-size: 1rem;
+    font-weight: 600;
+    line-height: 1.25rem;
+    text-decoration: none;
+}
```
Oh yeah, we don't need the wrapper element anymore, since all of the headers are going to start on the same line now. I'll simplify that 🙂

amaskiellEG replied:
Going to keep the height at 2.5rem and added bottom margin of .25rem, which looks just as good. All set!

nriserEG commented on README.md:
```diff
@@ -19,3 +19,24 @@ This global template replaces a hero banner with a new banner containing a backg
+3. Recipe
+    - ex. "Collaborative with Trending"
+4. Recommendations Attributes
+    - Show Product name
+    - Show Product description
+    - Show Product price
+    - Show Product rating
```
Make "Product" lowercase

nriserEG commented on product-recs/template.ts:
```diff
@@ -1,23 +1,60 @@
+
+    @header("Recommendations Attributes")
+
+    @title("Show Product name")
+    nameVisibility: boolean = true;
+
+    @title("Show Product description")
+    descriptionVisibility: boolean = true;
+
+    @title("Show Product price")
+    priceVisibility: boolean = true;
+
+    @title("Show Product rating")
+    ratingVisibility: boolean = false;
```
Make "Product" lowercase

amaskiellEG commented:
Think this needs to be re-approved, since I missed this comment: https://github.com/evergage/evergage-global-templates/pull/3#pullrequestreview-467338790

Just made all of those adjustments.

I just removed the `data-evg-product-` attributes, but also just remembered now why I had added them: they (or something similar `data-evg-item-id`) were in the recs example in the docs: https://developer.evergage.com/templates/campaign-stats. My hunch is that we might need to at least keep the `id` attribute for when support for PIKs is added?

avkieu commented:
@amaskiellEG to move this minor changes to a new branch, too many, now irrelevant, commits here to merge this.
